Elligator2 hash-to-curve for Bandersnatch

drskalman opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before hitting that submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

Implementation Elligator2 hash-to-curve parameters for Bandersnatch curve  according to IETF standard.

burdges commented on curves/ed_on_bls12_381_bandersnatch/Cargo.toml:
```diff
@@ -22,6 +22,7 @@ ark-relations = { version = "0.4.0", default-features = false }
 ark-serialize = { version = "0.4.0", default-features = false }
 ark-algebra-test-templates = { version = "0.4.0", default-features = false }
 ark-curve-constraint-tests = { path = "../curve-constraint-tests", default-features = false }
+sha2 = { version = "0.10", default-features = false }
```
Can we avoid propogating this sha2 xmd garbage?

burdges commented on curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs:
```diff
@@ -137,3 +138,55 @@ impl SWCurveConfig for BandersnatchConfig {
+    fn test_elligtor2_hash2curve_hashes_to_curve() {
+        let test_elligator2_to_curve_hasher = MapToCurveBasedHasher::<
+            Projective<BandersnatchConfig>,
+            DefaultFieldHasher<Sha512, 128>,
```
Again we should avoid sha2 xmd for new curves.

mmagician replied:
this is just a test though - what is the alternative here @burdges ?

burdges replied:
Yeah no worries.  We should make the shake h2f stuff work eventually though.

mmagician commented on curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs:
```diff
@@ -137,3 +138,55 @@ impl SWCurveConfig for BandersnatchConfig {
+    use sha2::Sha512;
+
+    #[test]
+    fn test_elligtor2_hash2curve_hashes_to_curve() {
```
I think for now it's ok to have these individual test. We should add some test templates though for hash-to-curve, I'll create an issue. Then we can have proper tests for Bandersnatch here, but also BLS12-377 & 381 on SWU.

drskalman commented on curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs:
```diff
@@ -137,3 +138,55 @@ impl SWCurveConfig for BandersnatchConfig {
 }
+
+// Elligator hash to curve Bandersnatch
+// sage: find_z_ell2(GF(52435875175126190479447740508185965837690552500527637822603658699938581184513))
```
https://github.com/arkworks-rs/algebra/pull/758/commits/c7d149386c8551ec687279535cbb48bab6e98258

mmagician replied:
Thank you!

mmagician commented on CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ## Pending
 
+- [\#XXX](https://github.com/arkworks-rs/algebra/pull/XXX) Implement Elligator2 hash-to-curve parameters for Bandersnatch.
```
```suggestion
- [\#XXX](https://github.com/arkworks-rs/algebra/pull/758) Implement Elligator2 hash-to-curve parameters for Bandersnatch.
```

mmagician commented:
Looks good. We can merge once the CHANGELOG is updated. It seems I can't commit to your branch directly.

drskalman commented:
> Looks good. We can merge once the CHANGELOG is updated. It seems I can't commit to your branch directly.

I had moved it to features before but I had forgotten  to push. I also added you as a contributor to the w3f fork for future. :pray:
